src/Time/Conversion.purs

```diff
@@ -0,0 +1,475 @@
+  let
+    startSlot = unwrap (unwrap start).slot
+    -- Round down to the nearest Slot to accept Milliseconds as input.
+    absSlot = startSlot + relSlot -- relative to system start
+    -- If `EraSummary` doesn't have an end, the condition is automatically
+    -- satisfied. We use `<=` as justified by the source code.
+    -- Potential FIXME: note the hack that we don't have `end` for the current
+    -- era, if we did not do this, there could be issues going far into the
+    -- future?
+    endSlot = maybe (absSlot + one) (unwrap <<< _.slot <<< unwrap) end
+  -- Check we are less than the end slot, or if equal, there is no excess:
+  unless (absSlot < endSlot || absSlot == endSlot && modTime == zero)
+    (throwError $ EndSlotLessThanSlotOrModNonZero (wrap absSlot) mt)
+  -- Potential FIXME: Do we want to use `safeZone` from `parameters`?
+  pure $ wrap absSlot
+
+-- | Get SlotLength in Milliseconds
+getSlotLength :: EraSummary -> BigInt
+getSlotLength (EraSummary { parameters }) =
+  unwrap (unwrap parameters).slotLength * factor
+
+--------------------------------------------------------------------------------
+
+-- | Converts a `POSIXTimeRange` to `SlotRange` given an `EraSummariesQR` and
+-- | `SystemStartQR` queried from Ogmios.
+posixTimeRangeToSlotRange
+  :: EraSummariesQR
+  -> SystemStartQR
+  -> POSIXTimeRange
+  -> Effect (Either PosixTimeToSlotError SlotRange)
+posixTimeRangeToSlotRange
+  eraSummaries
+  sysStart
+  (Interval { from: LowerBound s sInc, to: UpperBound e endInc }) = runExceptT
+  do
+    s' <- ExceptT $ convertBounds s
+    e' <- ExceptT $ convertBounds e
+    liftEither $ Right
+      $ Interval { from: LowerBound s' sInc, to: UpperBound e' endInc }
+  where
+  convertBounds
+    :: Extended POSIXTime
+    -> Effect (Either PosixTimeToSlotError (Extended Slot))
+  convertBounds (Finite pt) = posixTimeToSlot eraSummaries sysStart pt
+    <#> map Finite
+  convertBounds NegInf = pure $ Right NegInf
+  convertBounds PosInf = pure $ Right PosInf
+
+-- | Converts a `SlotRange` to `POSIXTimeRange` given an `EraSummariesQR` and
+-- | `SystemStartQR` queried from Ogmios.
+slotRangeToPosixTimeRange
+  :: EraSummariesQR
+  -> SystemStartQR
+  -> SlotRange
+  -> Effect (Either SlotToPosixTimeError POSIXTimeRange)
+slotRangeToPosixTimeRange
+  eraSummaries
+  sysStart
+  (Interval { from: LowerBound s sInc, to: UpperBound e endInc }) = runExceptT
+  do
+    s' <- ExceptT $ convertBounds s
+    e' <- ExceptT $ convertBounds e
+    liftEither $ Right
+      $ Interval { from: LowerBound s' sInc, to: UpperBound e' endInc }
+  where
+  convertBounds
+    :: Extended Slot
+    -> Effect (Either SlotToPosixTimeError (Extended POSIXTime))
+  convertBounds (Finite pt) = slotToPosixTime eraSummaries sysStart pt
+    <#> map Finite
+  convertBounds NegInf = pure $ Right NegInf
+  convertBounds PosInf = pure $ Right PosInf
+
+type TransactionValiditySlot =
+  { validityStartInterval :: Maybe Slot, timeToLive :: Maybe Slot }
```

Comment:
What do you mean by `timeToLive` here? in the function `slotRangeToTransactionValidity` that field is filled with the upper bound, but the field name almost seems to suggest a length?

How is `TransactionValiditySlot` used to build the transaction sent to the node?

Is this supposed to correspond to `ValidityInterval` from `cardano-ledger` ?
Note that describes an half-open interval: https://github.com/input-output-hk/cardano-ledger/blob/94b1ae3d6b66f4232f34060d89c2f12628998ef2/eras/shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Timelocks.hs#L100-L106